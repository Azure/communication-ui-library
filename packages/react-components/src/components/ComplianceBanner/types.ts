// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

/**
 * Subset of `CallCompositeStrings` needed by the ComplianceBanner component.
 * @internal
 */
export interface _ComplianceBannerStrings {
  close: string;
  complianceBannerNowOnlyRecording: string;
  complianceBannerNowOnlyTranscription: string;
  complianceBannerRecordingAndTranscriptionSaved: string;
  complianceBannerRecordingAndTranscriptionStarted: string;
  complianceBannerRecordingAndTranscriptionStopped: string;
  complianceBannerRecordingSaving: string;
  complianceBannerRecordingStarted: string;
  complianceBannerRecordingStopped: string;
  complianceBannerTranscriptionConsent: string;
  complianceBannerTranscriptionSaving: string;
  complianceBannerTranscriptionStarted: string;
  complianceBannerTranscriptionStopped: string;
  learnMore: string;
  privacyPolicy: string;
}

/**
 * @internal
 */
export type _ComplianceBannerProps = {
  callTranscribeState?: boolean;
  callRecordState?: boolean;
  strings: _ComplianceBannerStrings;
};
