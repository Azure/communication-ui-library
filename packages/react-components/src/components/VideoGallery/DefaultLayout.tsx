// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { Stack } from '@fluentui/react';
import React, { useMemo, useState, useRef } from 'react';
import { GridLayout } from '../GridLayout';
import { isNarrowWidth } from '../utils/responsive';
import { isShortHeight } from '../utils/responsive';
import { LayoutProps } from './Layout';
import { rootLayoutStyle } from './styles/DefaultLayout.styles';
import { videoGalleryLayoutGap } from './styles/Layout.styles';
import {
  MAX_GRID_PARTICIPANTS_NOT_LARGE_GALLERY,
  renderTiles,
  useOrganizedParticipants
} from './utils/videoGalleryLayoutUtils';
import { OverflowGallery } from './OverflowGallery';

/**
 * Props for {@link DefaultLayout}.
 *
 * @private
 */
export type DefaultLayoutProps = LayoutProps;

/**
 * DefaultLayout displays remote participants, local video component, and screen sharing component in
 * a grid an overflow gallery.
 *
 * @private
 */
export const DefaultLayout = (props: DefaultLayoutProps): JSX.Element => {
  const {
    remoteParticipants = [],
    localParticipant,
    dominantSpeakers,
    localVideoComponent,
    screenShareComponent,
    onRenderRemoteParticipant,
    styles,
    maxRemoteVideoStreams,
    parentWidth,
    parentHeight,
    pinnedParticipantUserIds = [],
    overflowGalleryPosition = 'horizontalBottom',
    spotlightedParticipantUserIds = []
  } = props;

  const isNarrow = parentWidth ? isNarrowWidth(parentWidth) : false;

  const isShort = parentHeight ? isShortHeight(parentHeight) : false;

  // This is for tracking the number of children in the first page of overflow gallery.
  // This number will be used for the maxOverflowGalleryDominantSpeakers when organizing the remote participants.
  // We need to add the local participant to the pinned participant count so we are placing the speakers correctly.
  const childrenPerPage = useRef(4);
  const remoteVideosOn = remoteParticipants.filter((p) => p.videoStream?.isAvailable).length > 0;
  const { gridParticipants, overflowGalleryParticipants } = useOrganizedParticipants({
    remoteParticipants,
    localParticipant,
    dominantSpeakers,
    maxGridParticipants: remoteVideosOn ? maxRemoteVideoStreams : MAX_GRID_PARTICIPANTS_NOT_LARGE_GALLERY,
    isScreenShareActive: !!screenShareComponent,
    maxOverflowGalleryDominantSpeakers: screenShareComponent
      ? childrenPerPage.current - ((pinnedParticipantUserIds.length + 1) % childrenPerPage.current)
      : childrenPerPage.current,
    pinnedParticipantUserIds,
    layout: 'default',
    spotlightedParticipantUserIds
  });

  /**
   * instantiate indexes available to render with indexes available that would be on first page
   *
   * For some components which do not strictly follow the order of the array, we might
   * re-render the initial tiles -> dispose them -> create new tiles, we need to take care of
   * this case when those components are here
   */
  const [indexesToRender, setIndexesToRender] = useState<number[]>([]);

  let { gridTiles, overflowGalleryTiles } = renderTiles(
    gridParticipants,
    onRenderRemoteParticipant,
    maxRemoteVideoStreams,
    indexesToRender,
    overflowGalleryParticipants,
    dominantSpeakers
  );

  if (localVideoComponent) {
    if (screenShareComponent || spotlightedParticipantUserIds.length > 0) {
      overflowGalleryTiles = [localVideoComponent].concat(overflowGalleryTiles);
    } else {
      gridTiles = [localVideoComponent].concat(gridTiles);
    }
  }

  const overflowGallery = useMemo(() => {
    if (overflowGalleryTiles.length === 0) {
      return null;
    }
    return (
      <OverflowGallery
        isNarrow={isNarrow}
        isShort={isShort}
        shouldFloatLocalVideo={false}
        overflowGalleryElements={overflowGalleryTiles}
        horizontalGalleryStyles={styles?.horizontalGallery}
        verticalGalleryStyles={styles?.verticalGallery}
        overflowGalleryPosition={overflowGalleryPosition}
        onFetchTilesToRender={setIndexesToRender}
        onChildrenPerPageChange={(n: number) => {
          childrenPerPage.current = n;
        }}
        parentWidth={parentWidth}
      />
    );
  }, [
    isNarrow,
    isShort,
    overflowGalleryTiles,
    styles?.horizontalGallery,
    overflowGalleryPosition,
    setIndexesToRender,
    styles?.verticalGallery,
    parentWidth
  ]);

  return (
    <Stack
      horizontal={overflowGalleryPosition === 'verticalRight'}
      styles={rootLayoutStyle}
      tokens={videoGalleryLayoutGap}
    >
      {props.overflowGalleryPosition === 'horizontalTop' ? overflowGallery : <></>}
      {screenShareComponent ? (
        screenShareComponent
      ) : (
        <GridLayout key="grid-layout" styles={styles?.gridLayout}>
          {gridTiles}
        </GridLayout>
      )}
      {overflowGalleryTrampoline(overflowGallery, props.overflowGalleryPosition)}
    </Stack>
  );
};

const overflowGalleryTrampoline = (
  gallery: JSX.Element | null,
  galleryPosition?: 'horizontalBottom' | 'verticalRight' | 'horizontalTop'
): JSX.Element | null => {
  return galleryPosition !== 'horizontalTop' ? gallery : <></>;
  return gallery;
};
