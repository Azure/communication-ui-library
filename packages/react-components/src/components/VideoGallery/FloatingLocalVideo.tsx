// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { ContextualMenu, IDragOptions, Stack } from '@fluentui/react';
import { _convertRemToPx } from '@internal/acs-ui-common';
import React, { useMemo } from 'react';
import { useTheme } from '../../theming';
import { _ICoordinates, _ModalClone } from '../ModalClone/ModalClone';
import {
  floatingLocalVideoModalStyle,
  floatinglocalVideoModalInitialPositionGapRem
} from './styles/FloatingLocalVideo.styles';
import { useLocale } from '../../localization';

const DRAG_OPTIONS: IDragOptions = {
  moveMenuItemText: 'Move',
  closeMenuItemText: 'Close',
  menu: ContextualMenu,
  keepInBounds: true
};

// Manually override the max position used to keep the modal in the bounds of its container.
// This is a workaround for: https://github.com/microsoft/fluentui/issues/20122
// Because our modal starts in the bottom right corner, we can say that this is the max (i.e. rightmost and bottomost)
// position the modal can be dragged to.
const modalMaxDragPosition = {
  x: _convertRemToPx(floatinglocalVideoModalInitialPositionGapRem),
  y: _convertRemToPx(floatinglocalVideoModalInitialPositionGapRem)
};

/**
 * @private
 */
export interface FloatingLocalVideoProps {
  // Local video component to make draggable
  localVideoComponent: JSX.Element;
  // Element id of layer host to constrain the dragging of local video
  layerHostId: string;
  // Parent component width in px
  parentWidth?: number;
  // Parent component height in px
  parentHeight?: number;
  // Local video width and height in rem
  localVideoSizeRem: {
    width: number;
    height: number;
  };
}

/**
 * @private
 */
export const FloatingLocalVideo = (props: FloatingLocalVideoProps): JSX.Element => {
  const { localVideoComponent, layerHostId, localVideoSizeRem, parentWidth, parentHeight } = props;

  const theme = useTheme();
  const strings = useLocale().strings.videoGallery;

  // The minimum drag position is the top left of the video gallery. i.e. the modal (PiP) should not be able
  // to be dragged offscreen and these are the top and left bounds of that calculation.
  const modalMinDragPosition: _ICoordinates | undefined = useMemo(
    () =>
      parentWidth && parentHeight
        ? {
            // We use -parentWidth/Height because our modal is positioned to start in the bottom right,
            // hence (0,0) is the bottom right of the video gallery.
            x:
              -parentWidth +
              _convertRemToPx(localVideoSizeRem.width) +
              _convertRemToPx(floatinglocalVideoModalInitialPositionGapRem),
            y:
              -parentHeight +
              _convertRemToPx(localVideoSizeRem.height) +
              _convertRemToPx(floatinglocalVideoModalInitialPositionGapRem)
          }
        : undefined,
    [parentHeight, parentWidth, localVideoSizeRem.width, localVideoSizeRem.height]
  );

  const modalStyles = useMemo(() => floatingLocalVideoModalStyle(theme, localVideoSizeRem), [theme, localVideoSizeRem]);
  const layerProps = useMemo(() => ({ hostId: layerHostId }), [layerHostId]);

  return (
    <_ModalClone
      isOpen={true}
      isModeless={true}
      dragOptions={DRAG_OPTIONS}
      styles={modalStyles}
      layerProps={layerProps}
      maxDragPosition={modalMaxDragPosition}
      minDragPosition={modalMinDragPosition}
      dataUiId="floating-local-video-host"
    >
      <Stack
        aria-label={strings.localVideoMovementAriaLabel ?? strings.localVideoMovementLabel}
        tabIndex={0}
        role={'dialog'}
        style={{ pointerEvents: 'auto' }}
      >
        {localVideoComponent}
      </Stack>
    </_ModalClone>
  );
};
