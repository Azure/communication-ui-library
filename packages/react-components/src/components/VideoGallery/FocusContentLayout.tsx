// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { LayerHost, Stack, mergeStyles } from '@fluentui/react';
import { LayoutProps } from './Layout';
import React from 'react';
import { innerLayoutStyle, layerHostStyle, rootLayoutStyle } from './styles/FloatingLocalVideoLayout.styles';
import { videoGalleryLayoutGap } from './styles/Layout.styles';
import { useId } from '@fluentui/react-hooks';

/**
 * Props for {@link FocusedContentLayout}.
 *
 * @private
 */
export interface FocusedContentLayoutProps extends LayoutProps {
  /**
   * Whether to display the local video camera switcher button
   */
  showCameraSwitcherInLocalPreview?: boolean;
  /**
   * Height of parent element
   */
  parentHeight?: number;
}

/**
 * Video gallery layout to focus on the screenshare stream that is present in the call.
 */
export const FocusedContentLayout = (props: FocusedContentLayoutProps): JSX.Element => {
  const { screenShareComponent } = props;

  const layerHostId = useId('layerhost');

  return (
    <Stack styles={rootLayoutStyle}>
      <LayerHost id={layerHostId} className={mergeStyles(layerHostStyle)} />
      <Stack styles={innerLayoutStyle} tokens={videoGalleryLayoutGap}>
        {screenShareComponent ? screenShareComponent : <></>}
      </Stack>
    </Stack>
  );
};
