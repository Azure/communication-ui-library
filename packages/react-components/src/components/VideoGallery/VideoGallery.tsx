// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { ContextualMenu, Icon, IDragOptions, Modal, Stack, Text, mergeStyles } from '@fluentui/react';
import React, { useMemo, useRef } from 'react';
import { smartDominantSpeakerParticipants } from '../../gallery';
import { useIdentifiers } from '../../identifiers/IdentifierProvider';
import { useLocale } from '../../localization';
import { useTheme } from '../../theming';
import {
  BaseCustomStylesProps,
  OnRenderAvatarCallback,
  VideoGalleryLocalParticipant,
  VideoGalleryRemoteParticipant,
  VideoStreamOptions
} from '../../types';
import { GridLayout } from '../GridLayout';
import { StreamMedia } from '../StreamMedia';
import {
  floatingLocalVideoModalStyle,
  floatingLocalVideoTileStyle,
  gridStyle,
  getHorizontalGalleryWrapperStyle,
  screenSharingContainer,
  screenSharingNotificationIconContainer,
  screenSharingNotificationIconStyle,
  screenSharingNotificationContainerCameraOnStyles,
  screenSharingNotificationContainerCameraOffStyles,
  screenSharingNotificationTextStyle,
  videoGalleryContainerStyle,
  videoGalleryOuterDivStyle
} from '../styles/VideoGallery.styles';
import { useContainerWidth, isNarrowWidth } from '../utils/responsive';
import { VideoTile, VideoTileStylesProps } from '../VideoTile';
import { HorizontalGallery } from './HorizontalGallery';
import { RemoteVideoTile } from './RemoteVideoTile';
import { getVideoTileOverrideColor } from '../utils/videoTileStylesUtils';

const emptyStyles = {};

/**
 * Strings of {@link VideoGalleryStrings} that can be overridden.
 *
 * @public
 */
export interface VideoGalleryStrings {
  /** Message to let user know they are sharing their screen. */
  screenSharingMessage: string;
}

/**
 * Props for {@link VideoGallery}.
 *
 * @public
 */
export interface VideoGalleryProps {
  /**
   * Allows users to pass an object containing custom CSS styles for the gallery container.
   *
   * @Example
   * ```
   * <VideoGallery styles={{ root: { border: 'solid 1px red' } }} />
   * ```
   */
  styles?: BaseCustomStylesProps;
  /** Layout of the video tiles. */
  layout?: 'default' | 'floatingLocalVideo';
  /** Local video particpant */
  localParticipant: VideoGalleryLocalParticipant;
  /** List of remote video particpants */
  remoteParticipants?: VideoGalleryRemoteParticipant[];
  /** List of dominant speaker userIds in the order of their dominance. 0th index is the most dominant. */
  dominantSpeakers?: Array<string>;
  /** Local video view options */
  localVideoViewOption?: VideoStreamOptions;
  /** Remote videos view options */
  remoteVideoViewOption?: VideoStreamOptions;
  /** Callback to create the local video stream view */
  onCreateLocalStreamView?: (options?: VideoStreamOptions) => Promise<void>;
  /** Callback to dispose of the local video stream view */
  onDisposeLocalStreamView?: () => void;
  /** Callback to render the local video tile*/
  onRenderLocalVideoTile?: (localParticipant: VideoGalleryLocalParticipant) => JSX.Element;
  /** Callback to create a remote video stream view */
  onCreateRemoteStreamView?: (userId: string, options?: VideoStreamOptions) => Promise<void>;
  /** Callback to render a remote video tile */
  onRenderRemoteVideoTile?: (remoteParticipant: VideoGalleryRemoteParticipant) => JSX.Element;
  onDisposeRemoteStreamView?: (userId: string) => Promise<void>;
  /** Callback to render a particpant avatar */
  onRenderAvatar?: OnRenderAvatarCallback;

  /**
   * Whether to display a mute icon beside the user's display name.
   * @defaultValue `true`
   */
  showMuteIndicator?: boolean;

  /**
   * Optional strings to override in component
   */
  strings?: Partial<VideoGalleryStrings>;
}

const DRAG_OPTIONS: IDragOptions = {
  moveMenuItemText: 'Move',
  closeMenuItemText: 'Close',
  menu: ContextualMenu,
  keepInBounds: true
};

/**
 * VideoGallery represents a {@link GridLayout} of video tiles for a specific call.
 * It displays a {@link VideoTile} for the local user as well as for each remote participants who joined the call.
 *
 * @public
 */
export const VideoGallery = (props: VideoGalleryProps): JSX.Element => {
  const {
    localParticipant,
    remoteParticipants,
    localVideoViewOption,
    remoteVideoViewOption,
    dominantSpeakers,
    onRenderLocalVideoTile,
    onRenderRemoteVideoTile,
    onCreateLocalStreamView,
    onCreateRemoteStreamView,
    onDisposeRemoteStreamView,
    styles,
    layout,
    onRenderAvatar,
    showMuteIndicator
  } = props;

  const ids = useIdentifiers();
  const theme = useTheme();
  const localeStrings = useLocale().strings.videoGallery;

  const shouldFloatLocalVideo = useMemo((): boolean => {
    return !!(layout === 'floatingLocalVideo' && remoteParticipants && remoteParticipants.length > 0);
  }, [layout, remoteParticipants]);

  const containerRef = useRef<HTMLDivElement>(null);
  const containerWidth = useContainerWidth(containerRef);
  const isNarrow = isNarrowWidth(containerWidth);
  const visibleVideoParticipants = useRef<VideoGalleryRemoteParticipant[]>([]);
  const visibleAudioParticipants = useRef<VideoGalleryRemoteParticipant[]>([]);

  visibleVideoParticipants.current = smartDominantSpeakerParticipants({
    participants: remoteParticipants?.filter((p) => p.videoStream?.isAvailable) ?? [],
    dominantSpeakers,
    visibleParticipants: visibleVideoParticipants.current.filter((p) => p.videoStream?.isAvailable)
  });

  // Create a map of visibleVideoParticipants for faster searching.
  // This map will be used to identify overflow participants. i.e., participants
  // that should be rendered in horizontal gallery.
  const visibleVideoParticipantsMap = {};
  visibleVideoParticipants.current.forEach((p) => {
    visibleVideoParticipantsMap[p.userId] = true;
  });
  // Max Tiles calculated inside that gallery can be passed to this function
  // to only return the max number of tiles that can be rendered in the gallery.
  visibleAudioParticipants.current = smartDominantSpeakerParticipants({
    participants: remoteParticipants?.filter((p) => !visibleVideoParticipantsMap[p.userId]) ?? [],
    dominantSpeakers,
    visibleParticipants: visibleAudioParticipants.current.filter((p) => !visibleVideoParticipantsMap[p.userId]),
    maxTiles: 100,
    maxDominantSpeakers: 6
  });

  // If there are no video participants, we assign all audio participants as grid participants and assign
  // an empty array as horizontal gallery partipants to avoid rendering the horizontal gallery.
  const gridParticipants =
    visibleVideoParticipants.current.length > 0 ? visibleVideoParticipants.current : visibleAudioParticipants.current;
  const horizontalGalleryParticipants =
    visibleVideoParticipants.current.length > 0 ? visibleAudioParticipants.current : [];

  const screenSharingNotification = useMemo((): JSX.Element | undefined => {
    if (!localParticipant.isScreenSharingOn) {
      return undefined;
    }

    const screenSharingNotificationContainerStyle = mergeStyles(
      localParticipant.videoStream?.renderElement
        ? screenSharingNotificationContainerCameraOnStyles
        : screenSharingNotificationContainerCameraOffStyles,
      getVideoTileOverrideColor(!!localParticipant.videoStream?.renderElement, theme, 'neutralSecondary')
    );

    const screenSharingNotificationIconThemedStyle = mergeStyles(
      screenSharingNotificationIconStyle,
      getVideoTileOverrideColor(!!localParticipant.videoStream?.renderElement, theme, 'neutralTertiary')
    );

    return (
      <Stack horizontalAlign={'center'} verticalAlign={'center'} className={screenSharingContainer}>
        <Stack
          horizontalAlign={'center'}
          verticalAlign={'center'}
          className={screenSharingNotificationContainerStyle}
          tokens={{ childrenGap: '1rem' }}
        >
          <Stack horizontal verticalAlign={'center'} className={screenSharingNotificationIconContainer}>
            <Icon iconName="ControlButtonScreenShareStart" className={screenSharingNotificationIconThemedStyle} />
          </Stack>
          <Text className={screenSharingNotificationTextStyle} aria-live={'polite'}>
            {props.strings?.screenSharingMessage ?? localeStrings.screenSharingMessage}
          </Text>
        </Stack>
      </Stack>
    );
  }, [
    localParticipant.isScreenSharingOn,
    localParticipant.videoStream,
    localParticipant.videoStream?.renderElement,
    theme
  ]);

  /**
   * Utility function for memoized rendering of LocalParticipant.
   */
  const defaultOnRenderLocalVideoTile = useMemo((): JSX.Element => {
    const localVideoStream = localParticipant?.videoStream;

    if (onRenderLocalVideoTile) return onRenderLocalVideoTile(localParticipant);

    let localVideoTileStyles: VideoTileStylesProps = {};
    if (shouldFloatLocalVideo) {
      localVideoTileStyles = floatingLocalVideoTileStyle;
    }

    if (localVideoStream && !localVideoStream.renderElement) {
      onCreateLocalStreamView && onCreateLocalStreamView(localVideoViewOption);
    }

    return (
      <VideoTile
        userId={localParticipant.userId}
        renderElement={
          localVideoStream?.renderElement ? (
            <StreamMedia videoStreamElement={localVideoStream.renderElement} />
          ) : undefined
        }
        displayName={localParticipant?.displayName}
        styles={localVideoTileStyles}
        onRenderPlaceholder={localParticipant.isScreenSharingOn ? () => <></> : onRenderAvatar}
        isMuted={localParticipant.isMuted}
        showMuteIndicator={showMuteIndicator}
      >
        {screenSharingNotification}
      </VideoTile>
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    localParticipant,
    localParticipant.isScreenSharingOn,
    localParticipant.videoStream,
    localParticipant.videoStream?.renderElement,
    onCreateLocalStreamView,
    onRenderLocalVideoTile,
    onRenderAvatar,
    shouldFloatLocalVideo
  ]);

  /**
   * Utility function for memoized rendering of RemoteParticipants.
   */
  const remoteGridTiles = useMemo(() => {
    // If user provided a custom onRender function return that function.
    if (onRenderRemoteVideoTile) {
      return remoteParticipants?.map((participant) => onRenderRemoteVideoTile(participant));
    }

    // Else return Remote Stream Video Tiles
    return gridParticipants?.map((participant): JSX.Element => {
      const remoteVideoStream = participant.videoStream;
      return (
        <RemoteVideoTile
          key={participant.userId}
          userId={participant.userId}
          onCreateRemoteStreamView={onCreateRemoteStreamView}
          onDisposeRemoteStreamView={onDisposeRemoteStreamView}
          isAvailable={remoteVideoStream?.isAvailable}
          isMuted={participant.isMuted}
          isSpeaking={participant.isSpeaking}
          renderElement={remoteVideoStream?.renderElement}
          displayName={participant.displayName}
          remoteVideoViewOption={remoteVideoViewOption}
          onRenderAvatar={onRenderAvatar}
          showMuteIndicator={showMuteIndicator}
        />
      );
    });
  }, [
    onRenderRemoteVideoTile,
    gridParticipants,
    remoteParticipants,
    onCreateRemoteStreamView,
    onDisposeRemoteStreamView,
    remoteVideoViewOption,
    onRenderAvatar,
    showMuteIndicator
  ]);

  const gridTiles = remoteGridTiles ?? [];
  if (!shouldFloatLocalVideo && localParticipant) {
    gridTiles.push(
      <Stack data-ui-id={ids.videoGallery} horizontalAlign="center" verticalAlign="center" className={gridStyle} grow>
        {localParticipant && defaultOnRenderLocalVideoTile}
      </Stack>
    );
  }

  return (
    <div ref={containerRef} className={videoGalleryOuterDivStyle}>
      <Stack id={floatingTileHostId} grow styles={videoGalleryContainerStyle}>
        {shouldFloatLocalVideo && (
          <Modal
            isOpen={true}
            isModeless={true}
            dragOptions={DRAG_OPTIONS}
            styles={floatingLocalVideoModalStyle(isNarrow)}
            layerProps={{ hostId: floatingTileHostId }}
          >
            {localParticipant && defaultOnRenderLocalVideoTile}
          </Modal>
        )}
        <GridLayout styles={styles ?? emptyStyles}>{gridTiles}</GridLayout>
        {horizontalGalleryParticipants && horizontalGalleryParticipants.length > 0 && (
          <Stack style={getHorizontalGalleryWrapperStyle(isNarrow)}>
            <HorizontalGallery
              onCreateRemoteStreamView={onCreateRemoteStreamView}
              onDisposeRemoteStreamView={onDisposeRemoteStreamView}
              onRenderAvatar={onRenderAvatar}
              onRenderRemoteVideoTile={onRenderRemoteVideoTile}
              participants={horizontalGalleryParticipants}
              remoteVideoViewOption={remoteVideoViewOption}
              showMuteIndicator={showMuteIndicator}
              hideRemoteVideoStream={shouldFloatLocalVideo}
              rightGutter={shouldFloatLocalVideo ? (isNarrow ? 64 : 176) : undefined} // to leave a gap for the floating local video
            />
          </Stack>
        )}
      </Stack>
    </div>
  );
};

const floatingTileHostId = 'UILibraryFloatingTileHost';
