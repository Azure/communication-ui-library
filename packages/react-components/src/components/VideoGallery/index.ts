// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

export { VideoGallery } from './VideoGallery';
export type { VideoGalleryProps, VideoGalleryStrings } from './VideoGallery';
