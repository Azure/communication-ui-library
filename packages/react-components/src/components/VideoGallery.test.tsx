// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { initializeIcons, Stack } from '@fluentui/react';
import Enzyme, { mount, ReactWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { VideoGalleryLocalParticipant } from '../types';
import { VideoGallery } from './VideoGallery';

// create a document div with purple background
const mockLocalRenderElement = document.createElement('div');
mockLocalRenderElement.style.height = '50px';
mockLocalRenderElement.style.width = '50px';
mockLocalRenderElement.style.backgroundColor = 'purple';
mockLocalRenderElement.innerHTML = '<div>LOCAL PARTICIPANT</div>';

const mockLocalParticipant: VideoGalleryLocalParticipant = {
  userId: 'localParticipant',
  isMuted: false,
  displayName: 'localParticipant',
  videoStream: {
    id: 1,
    isAvailable: true,
    isReceiving: true,
    isMirrored: true,
    renderElement: mockLocalRenderElement
  },
  isScreenSharingOn: false
};

describe('VideoGallery', () => {
  beforeAll(() => {
    Enzyme.configure({ adapter: new Adapter() });
    initializeIcons();
  });

  test('should render local video tile as the only tile', () => {
    const root = mountVideoGalleryWithDefaults();
    // create jest snapshot
    expect(root.render()).toMatchSnapshot('VideoGallery-with-local-participant');
  });
});

const mountVideoGalleryWithDefaults = (): ReactWrapper => {
  let root;
  act(() => {
    root = mount(
      <Stack style={{ height: '100px', width: '100px', position: 'relative' }}>
        <VideoGallery localParticipant={mockLocalParticipant} />
      </Stack>
    );
  });
  return root;
};
