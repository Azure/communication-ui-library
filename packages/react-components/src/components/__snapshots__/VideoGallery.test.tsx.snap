// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoGallery should render local video tile as the only tile: VideoGallery-with-local-participant 1`] = `
LoadedCheerio {
  "0": <div
    class="ms-Stack css-119"
    style="height: 100px; width: 100px; position: relative;"
  >
    <div
      class="css-114"
      data-ui-id="video-gallery"
    >
      <div
        class="ms-Stack css-121"
      >
        <div
          class="css-123"
        >
          <div
            aria-label="Movable Local Video Tile"
            class="ms-Stack css-119"
            role="dialog"
            tabindex="0"
          >
            <div
              class="ms-Stack css-129"
              data-ui-id="video-tile"
            >
              <div
                class="ms-Stack css-130"
              >
                <div
                  class="ms-Stack css-131"
                />
                <div
                  class="css-132"
                >
                  <div
                    class="css-133"
                    data-ui-id="stream-media-container"
                  >
                    <div
                      style="height: 50px; width: 50px; background-color: purple;"
                    >
                      <div>
                        LOCAL PARTICIPANT
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="ms-Stack css-134"
              >
                <div
                  class="ms-Stack css-135"
                >
                  <span
                    class="css-136"
                    title="You"
                  >
                    You
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ms-LayerHost css-120"
          id="layerhost0"
        />
      </div>
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
